### PR TITLE
Fix truncated audit-log when using DynamoDB

### PR DESCRIPTION
**Pull request by pierrebeaucamp**
This is a fix for #4977. Teleport will continue to query DynamoDB until
the response doesn't contain a `LastEvaluatedKey` anymore, which signals
the end of the result set.

### CODE REVIEW

**Review by alex-kovoy — APPROVED** [review]

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: lib/events/dynamoevents/dynamoevents.go
Comment (alex-kovoy) [reply lib/events/dynamoevents/dynamoevents.go:506]: Lets use some memory limit and not the number of iterations. DynamoDB pagination is based on the size of query result (max 900kb) meaning that  900kb may have just a few events in theory. So 300 iteration potentially can use 300mb of RAM per request. @r0mant @awly is there a way to determine the RAM available and take % from it as a max buffer size during iteration? (this does not sound as reliable solution of course but what are other options until we add streaming?)



Since we do not have streaming 



We have a hard limit of 10k events (the limit value). 

DynamoDB pagination is based on size (not the number of events), so if size is too greate it would consume all memory regardless of this

Comment (klizhentas) [reply lib/events/dynamoevents/dynamoevents.go:506]: @alex-kovoy having dynamic memory computation is overly complex, let's not do that. I'm OK to set some large hardcoded value of 100MB per result set though instead of the number of iterations, this is a good idea.

Comment (alex-kovoy) [reply lib/events/dynamoevents/dynamoevents.go:506]: We can add some queue and place incoming requests on hold while the queue is being processed. Pending requests will have some timeout values.

Comment (klizhentas) [reply lib/events/dynamoevents/dynamoevents.go:506]: No need for queue and processing of incoming requests. If we want to make it right, we should expose iterator to the higher level API to allow for efficient pagination.

Comment (alex-kovoy) [reply lib/events/dynamoevents/dynamoevents.go:506]: When hit with several requests per second, this 100mb limit may lead to DoS.

Comment (a-palchikov) [reply lib/events/dynamoevents/dynamoevents.go:506]: I might also add that a potentially unbounded loop should make progress. The primary key (`lastEvaluatedKey`) seems a likely candidate for progression which should change between two sequential invocations unless it is empty.

File: lib/events/dynamoevents/dynamoevents.go
Comment (pierrebeaucamp) [reply lib/events/dynamoevents/dynamoevents.go]: I have a quite trivial test ready, but it increases the 40s runtime to multiple minutes (on my machine at least - us-west-1 is not very close). I'm reluctant to add such a long running test. Let me know if you think that's ok

Comment (a-palchikov) [reply lib/events/dynamoevents/dynamoevents.go]: I don't think a longer test should be an issue. I know, for example, other companies like Google have been grouping the tests by time complexity. Going that route, we can also choose which tests to run during development and have a way of switching on the "heavier" one with the heavy tests being disabled by default.

Comment (pierrebeaucamp) [reply lib/events/dynamoevents/dynamoevents.go]: Alright, it's hidden behind the `TEST_AWS` flag anyway, so hopefully the impact isn't too big

Comment (webvictim) [reply lib/events/dynamoevents/dynamoevents.go]: That's fine, we don't run those tests by default.

**Comment by alex-kovoy** [comment]
@klizhentas @awly @r0mant please review guys!

File: lib/events/dynamoevents/dynamoevents.go
Comment (r0mant) [inline lib/events/dynamoevents/dynamoevents.go:548]: A comment explaining that empty LastEvaluatedKey means the last page of results was returned wouldn't hurt (I had to go to the docs to confirm).

**Review by r0mant — COMMENTED** [review]
I'm also curious what is the maximum number of events this can return now? Max 300 iterations * default return limit = ?

File: lib/events/dynamoevents/dynamoevents.go
Comment (awly) [reply lib/events/dynamoevents/dynamoevents.go:554]: yeah, log an error and `return values, nil`
it's better to degrade gracefully - return as many entries as possible when the backend accumulates too much, instead of showing errors to the user

File: lib/events/dynamoevents/dynamoevents.go
Comment (awly) [inline lib/events/dynamoevents/dynamoevents.go:506]: define a new constant instead of using a multiple of something unrelated

File: lib/events/dynamoevents/dynamoevents.go
Comment (pierrebeaucamp) [reply lib/events/dynamoevents/dynamoevents.go:554]: I was trying to follow existing Teleport conventions, for example: https://github.com/gravitational/teleport/blob/master/lib/backend/dynamo/dynamodbbk.go#L401

File: lib/events/dynamoevents/dynamoevents.go
Comment (pierrebeaucamp) [reply lib/events/dynamoevents/dynamoevents.go:506]: Again, just trying to follow precedent: https://github.com/gravitational/teleport/blob/master/lib/backend/dynamo/dynamodbbk.go#L389

This was also suggested by Sasha in an earlier comment. Personally I agree though, since I don't know the background behind `DefaultLargeLimit` and why it should be divided by 100 in these cases.

Comment (alex-kovoy) [reply lib/events/dynamoevents/dynamoevents.go:506]: @pierrebeaucamp I think we agreed to use hardcoded 100mb value for the buffer size (suggested by Sasha) per request instead of `DefaultLargeLimit`. 

```go
for downloaded < maxBufferSize  {

    // 900mb is dynamo db max page size
    downloaded += 900; 
}
```

Comment (awly) [reply lib/events/dynamoevents/dynamoevents.go:506]: Not all existing code is good code that should be followed.
Let's do what makes the most sense for the problem at hand.

The doc comment on `DefaultLargeLimit` is:
```
	// DefaultLargeLimit is used to specify some very large limit when limit is not specified
	// explicitly to prevent OOM
```
which is not very meaningful and can't possibly be the right value for all types of "limit"

File: lib/events/dynamoevents/dynamoevents.go
Comment (awly) [reply lib/events/dynamoevents/dynamoevents.go:554]: IMO the code you linked is wrong too

File: lib/events/dynamoevents/dynamoevents.go
Comment (a-palchikov) [inline lib/events/dynamoevents/dynamoevents.go:508]:
```suggestion
	// DynamoDB responses are capped at 900KB according to their documentation.
```

File: lib/events/dynamoevents/dynamoevents.go
Comment (a-palchikov) [inline lib/events/dynamoevents/dynamoevents.go:509]: which units is this? If bytes, this needs to be `100*1024*1024` or `100<<20` for 100MiB

Comment (alex-kovoy) [reply lib/events/dynamoevents/dynamoevents.go:509]: @pierrebeaucamp could you move it to a const (the same file) and call it something like:

```
pageSizeKbs = 900
maxResponseSizeKbs = 100 * pageSizeKbs
```

File: lib/events/dynamoevents/dynamoevents.go
Comment (a-palchikov) [inline lib/events/dynamoevents/dynamoevents.go:164]: nit: it would be more appropriate to link to the documentation instead

File: lib/events/dynamoevents/dynamoevents.go
Comment (a-palchikov) [inline lib/events/dynamoevents/dynamoevents.go:514]: And it's not just going from `0` to `99` because it's more descriptive that way? Sorry if I missed that comment.

**Review by a-palchikov — APPROVED** [review]

File: lib/events/dynamoevents/dynamoevents.go
Comment (awly) [reply lib/events/dynamoevents/dynamoevents.go:514]: Yeah, this is basically `for i := 0; i < 100; i++` with extra steps.
You can just leave a comment saying that 100 maxed-out responses from dynamo here will eat ~100MB

**Review by awly — APPROVED** [review]

File: lib/events/dynamoevents/dynamoevents.go
Comment (r0mant) [inline lib/events/dynamoevents/dynamoevents.go:514]: Sorry, I may be missing something but this condition is pretty confusing to me. The idea of limiting by memory makes sense, however I don't see how it's actually enforced here - i.e. the `responseSize` is not the actual size of the response but basically just a "weighted" iteration number.

So how is this different from, say, just `for i := 0; i < 100; i += 1`? If the intent is to convey the fact that limiting number of iterations to a 100 will limit max returned response size to 90MB, then why not just put that in the comment?

Comment (alex-kovoy) [reply lib/events/dynamoevents/dynamoevents.go]: You are right as we are not counting the actual size it makes sense to use a basic loop. I've addressed the changes.

**Review by r0mant — APPROVED** [review]

**Comment by webvictim** [comment]
Please don't use administrator privileges to override changes being requested on the PR. The original reviewer who requested the changes should approve the updates before merging.

**Pull request merged**
